### PR TITLE
Nested Groups in sideNav

### PR DESCRIPTION
**Pull request by qtzar**
This PR adds a new way to generate the side nav that accounts for nested groups in the structurizer DSL as the softwareSystem level.

I have tested this with a workspace.dsl that contains nested groups as well as the example workspace.dsl that does not contain any groups.

### CODE REVIEW

**Comment by qtzar** [comment]
Screenshot of default workspace with no groups.

Screenshot 2023-03-17 at 3 37 44 PM

Screenshot with groups added.

Screenshot 2023-03-17 at 3 42 03 PM

**Comment by dirkgroot** [comment]
@qtzar Thanks for your PR. I'm not sure if I like this change, for a couple of reasons:

- User experience:
  - It add visual clutter to the menu, because of the dashed lines and `+` and `-` symbols.
  - The style of the menu is not consistent with the style of the rest of the menu. It's more crammed, and indentation is different.
  - All groups are collapsed by default, which (especially for smaller models) isn't really necessary, I think.
  - Navigation through the available systems via the menu now requires more mouse clicks.
- I'm not sure if this is the best way to visualize groups in the site. I think it's a nice addition, but I think I'd prefer a somewhat more minimalistic approach, but I'm not sure what that should look like. Any ideas (CC @jp7677)?

**Comment by qtzar** [comment]
That was an older screenshot, I did a little extra tidy up in a later commit.

The style of the menu is now consistent with the rest of the menu and now supports the is-active class. I've also updated the js to auto-expand the menu to the correct level when looking at a nested softwareSystem.

While navigation though the available systems does seem to have a lot more clicks, in a large organization with 400+ software systems the side menu was way too big to list them as a flat structure. Now with them nested it is much easier to manage for a person to drop down to the correct group. Also with this PR combined with the navigation links I added to container and component views a user can very quickly move around the system.

Screenshot 2023-03-24 at 12 02 41 PM

**Comment by qtzar** [comment]
@dirkgroot Just pushed an update that should make this even better.

You can now add `--properties sitegen.properties` to the command line to set a property which will enable or disable the nested groups in the side nav.

siteGen.nestGroupsInNav = <true/false>

The site generator will default this to false if the properties file is not specified, not found or if that particular value is not set in the properties file.

**Comment by jp7677** [comment]
I'm torn about introducing a tree view for the software systems. Like `dirkgroot` I'm also found of the simple flat list of all software systems. Considering that groups allow a lot of flexibility (multiple groups, nesting) and can be used for a lot of different meanings, I'm in doubt that we can find a presentation that fits everyone.

My advise for how to present (certain) groups more prominently would be to use global documentation (which appear in the menu, like `System Landscape` in the example) en then use embedded views with exactly the groups and systems plus context you need for a certain use case.

**Comment by jp7677** [comment]
I'm torn about introducing a tree view for the software systems. Like `dirkgroot` I'm also found of the simple flat list of all software systems. Considering that groups allow a lot of flexibility (multiple groups, nesting) and can be used for a lot of different meanings, I'm in doubt that we can find a presentation that fits everyone.

My advise for how to present (certain) groups more prominently would be to use global documentation (which appear in the menu, like `System Landscape` in the example) en then use embedded views with exactly the groups and systems plus context you need for a certain use case.

**Comment by qtzar** [comment]
@jp7677 Based on the comments from @dirkgroot I introduced the ability to enable/disable it via a properties file. I think this gives the best of both worlds. It also give a good start on issue #121 as a side effect.

**Comment by pietermees** [comment]
@qtzar we'd love to use this functionality, I was wondering if this was still something you were pursuing or whether it's stale?

**Comment by qtzar** [comment]
I would LOVE to have this functionality merged in also. I probably have to rebuild it a little as the main branch has moved on quite a bit and I'd make sure that it can be enabled/disabled as wanted.

@dirkgroot @jp7677 What do you think? Could this feature be considered again?

**Comment by qtzar** [comment]
I rebuilt the functionality on top of the most current main branch and then forced a commit to my branch to eliminate all previous commits.

**Comment by qtzar** [comment]
Here is an updated screenshot ( ignore the blacked out bit to hide company name ). As mentioned in one of the outdated comments we have 400+ systems and the simple flat list is just unusable. We do have sets of landscape views based on certain major groups and show sub groups within those landscape docs but it doesn't negate that fact that 400+ systems means a really long scroll on the page.

Screenshot 2023-09-21 at 1 04 51 PM

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Page.kt [hunk 1]
```diff
@@ -27,7 +27,14 @@ private fun HTML.headFragment(viewModel: PageViewModel) {
             href = "./" + "/style-branding.css".asUrlToFile(viewModel.url)
         )
 
-        if (viewModel.includeAdmonition) 
+        if(viewModel.includeTreeview){
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Page.kt:30]: missing space

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Page.kt [hunk 2]
```diff
@@ -64,6 +71,19 @@ private fun HTML.bodyFragment(viewModel: PageViewModel, block: DIV.() -> Unit) {
             updateSiteErrorHero()
         if (viewModel.includeAdmonition)
             markdownAdmonitionScript(viewModel)
+        if (viewModel.includeTreeview)
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Page.kt:74]: This line looks like a typo

**Review by jp7677 — CHANGES_REQUESTED** [review]
Sorry for the late review, it has been rather busy :(.

This PR is shaping up nicely and since it is an opt-in, I'm fine to get this in. Could you please look at my comments/suggestions. I haven't tested them, so likely bugs are included ;)

Could you please also add a few tests that validate the construction of the MenuNodes?

File: README.md [hunk 3]
```diff
@@ -261,9 +261,10 @@ architecture model:
...
-| `generatr.svglink.target`               | Specifies the link target for element links in the exported svg                                                                                                                                                                                                                                                                                  | `_top`    | `_self`                                              |
+| `generatr.search.language`              | Indexing/stemming language for the search index. See [Lunr language support](https://github.com/olivernn/lunr-languages)                                                                                                                                                             | `en`      | `nl`                |
```
Comment (jp7677) [inline README.md:264]: Could you please check the formatting here?

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MenuViewModel.kt [hunk 4]
```diff
@@ -27,4 +27,35 @@ class MenuViewModel(generatorContext: GeneratorContext, private val pageViewMode
...
+
+        return rootNode.toMenuNode()
+    }
+
+    private val softwareSystemPaths = generatorContext.workspace.model.includedSoftwareSystems
+        .map { it.group + "/" + it.name }
+        .sortedBy { it.lowercase() }
+
+    private val delimiter = '/'
+
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MenuViewModel.kt:60]: Not needed empty line.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Menu.kt [hunk 5]
```diff
@@ -30,3 +38,43 @@ private fun ASIDE.menuItemLinks(items: List<LinkViewModel>) {
...
+            }
+
+            ul(classes = "listree-submenu-items") {
+                for (child in node.children) {
+                    buildHtmlTree(child,viewModel).invoke(this)
+                }
+            }
+        }
+    }
+    if (node.name.isEmpty() && node.children.isNotEmpty()) {
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Menu.kt:72]: Empty line above please.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Menu.kt [hunk 6]
```diff
@@ -30,3 +38,43 @@ private fun ASIDE.menuItemLinks(items: List<LinkViewModel>) {
...
+        ul(classes = "listree-submenu-items") {
+            for (child in node.children) {
+                buildHtmlTree(child, viewModel).invoke(this)
+            }
+        }
+    }
+
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/Menu.kt:79]: Not needed empty line.

File: src/test/kotlin/nl/avisi/structurizr/site/generatr/site/model/MenuViewModelTest.kt [hunk 7]
```diff
@@ -136,6 +136,27 @@ class MenuViewModelTest : ViewModelTest() {
...
+
+        MenuViewModel(generatorContext, createPageViewModel(generatorContext, url = HomePageViewModel.url()))
+            .let {
+                assertThat(it.softwareSystemNodes().children).hasSize(2)
+                assertThat(it.softwareSystemNodes().children[0].name).isEqualTo("Group 1")
+                assertThat(it.softwareSystemNodes().children[0].children).hasSize(3)
+                assertThat(it.softwareSystemNodes().children[0].children[0].name).isEqualTo("Group 3")
+                assertThat(it.softwareSystemNodes().children[0].children[0].children[0].name).isEqualTo("System 4")
+            }
+
```
Comment (jp7677) [inline src/test/kotlin/nl/avisi/structurizr/site/generatr/site/model/MenuViewModelTest.kt:156]: Not needed empty lines.

**Review by jp7677 — CHANGES_REQUESTED** [review]
Just some whitespace nits left...

**Comment by qtzar** [comment]
@jp7677 Thanks for taking the time to help with this PR.

**Review by jp7677 — APPROVED** [review]

**Comment by jp7677** [comment]
@qtzar All good, but unfortunately conflicts, can you please bring this branch in line with `main`?

**Comment by qtzar** [comment]
@jp7677 I'm not showing my branch being behind `main`.

**Pull request merged**